File: Baselines/dComBat/distributedCombat_comparison.py
import pandas as pd
import numpy as np
from rpy2.robjects import r, numpy2ri, pandas2ri
import neuroCombat as nc
import distributedCombat as dc


numpy2ri.activate()
pandas2ri.activate()
r(
    """
library(matrixStats)
# Simulate data
set.seed(8888)

p=10000 # Number of features
q=3 # Number of covariates
n=100
batch = rep(1:4, n/4) #Batch variable for the scanner id
batch <- as.factor(batch)

mod = matrix(runif(q*n), n, q) #Random design matrix
dat = matrix(runif(p*n), p, n) #Random data matrix
"""
)

dat = pd.DataFrame(r["dat"])
mod = pd.DataFrame(r["mod"])
batch_col = "batch"
covars = pd.DataFrame({batch_col: r["batch"]})

#### Ground truth ComBat outputs ####
com_out = nc.neuroCombat(dat, covars, batch_col)
com_out_ref = nc.neuroCombat(dat, covars, batch_col, ref_batch="1")

#### Distributed ComBat: No reference batch ####
### Step 1
site_outs = []
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(df, bat, x, verbose=True, file=f)
    site_outs.append(f)

central = dc.distributedCombat_central(site_outs)


### Step 2
site_outs = []
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(
        df, bat, x, verbose=True, central_out=central, file=f
    )
    site_outs.append(f)

central = dc.distributedCombat_central(site_outs)

### Compare distributed vs original
site_outs = []
error = []
perror = []  # percent difference
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(df, bat, x, central_out=central, file=f)
    site_outs.append(f)
    print(
        "ground truth", np.allclose(out["dat_combat"], com_out["data"][:, s], atol=0.1)
    )
    error.append(com_out["data"][:, s] - out["dat_combat"])
    perror.append(
        abs(com_out["data"][:, s] - out["dat_combat"]) / com_out["data"][:, s]
    )


print("error", np.array(error))
print("perror", np.array(perror))


print("===============================================================")


#### Distributed ComBat: With reference batch ####
### Step 1
site_outs = []
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(df, bat, x, verbose=True, file=f, ref_batch="1")
    site_outs.append(f)

central = dc.distributedCombat_central(site_outs, ref_batch="1")

### Step 2
site_outs = []
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(
        df, bat, x, verbose=True, central_out=central, file=f, ref_batch="1"
    )
    site_outs.append(f)

central = dc.distributedCombat_central(site_outs, ref_batch="1")

### Compare distributed vs original
site_outs = []
error = []
perror = []  # percent difference
for b in covars[batch_col].unique():
    s = list(map(lambda x: x == b, covars[batch_col]))
    df = dat.loc[:, s]
    bat = covars[batch_col][s]
    x = mod.loc[s, :]
    f = "site_out_" + str(b) + ".pickle"
    out = dc.distributedCombat_site(
        df, bat, x, central_out=central, file=f, ref_batch="1"
    )
    site_outs.append(f)
    print(
        "ground truth", np.allclose(out["dat_combat"], com_out["data"][:, s], atol=0.5)
    )
    error.append(com_out["data"][:, s] - out["dat_combat"])
    perror.append(
        abs(com_out["data"][:, s] - out["dat_combat"]) / com_out["data"][:, s]
    )

print("error", np.array(error))
print("perror", np.array(perror))

File: Baselines/dComBat/Comparison with ComBat/distributedCombat_comparison.R
library(neuroCombat)
library(matrixStats)
source("../distributedCombat.R")
source("../neuroComBat_helpers.R")
source("../neuroComBat.R")

# Simulate data
set.seed(8888)

p=10000 # Number of features
q=3 # Number of covariates
n=100
batch = rep(1:4, n/4) #Batch variable for the scanner id
batch <- as.factor(batch)

mod = matrix(runif(q*n), n, q) #Random design matrix
dat = matrix(runif(p*n), p, n) #Random data matrix

#### Ground truth ComBat outputs ####
com_out <- neuroCombat(dat, batch, mod)
com_out_ref <- neuroCombat(dat, batch, mod, ref.batch = "1")

#### Distributed ComBat: No reference batch ####
### Step 1
site.outs <- NULL
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]
  
  site.outs <- c(site.outs, list(distributedCombat_site(df, bat, x)))
}
central <- distributedCombat_central(site.outs)

### Step 2
site.outs <- NULL
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]
  
  site.outs <- c(site.outs, list(distributedCombat_site(df, bat, x,
                                                        central.out = central)))
}
central <- distributedCombat_central(site.outs)

### Compare distributed vs original
site.outs <- NULL
error <- NULL
perror <- NULL # percent difference
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]

  site.out <- distributedCombat_site(df, bat, x, central.out = central)
  site.outs <- c(site.outs, site.out)
  
  error <- c(error, max(c(com_out$dat.combat[,s] - site.out$dat.combat)))
  perror <- c(perror,
              max(c(abs(com_out$dat.combat[,s] - site.out$dat.combat)/
                      site.out$dat.combat)))
}

error # Maximum difference by batch
perror # Maximum difference as percent by batch

#### Distributed ComBat: With reference batch ####
### Step 1
site.outs <- NULL
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]
  
  site.outs <- c(site.outs, list(distributedCombat_site(df, bat, x, ref.batch = "1")))
}
central <- distributedCombat_central(site.outs, ref.batch = "1")

### Step 2
site.outs <- NULL
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]
  
  site.outs <- c(site.outs, list(distributedCombat_site(df, bat, x, ref.batch = "1",
                                                        central.out = central)))
}
central <- distributedCombat_central(site.outs, ref.batch = "1")

### Compare distributed vs original
site.outs <- NULL
error_ref <- NULL
perror_ref <- NULL # percent difference
for (b in unique(batch)) {
  s <- batch == b
  df <- dat[,s]
  bat <- batch[s]
  x <- mod[s,]
  
  site.out <- distributedCombat_site(df, bat, x, ref.batch = "1", 
                                     central.out = central)
  site.outs <- c(site.outs, site.out)
  
  error_ref <- c(error_ref, max(c(com_out_ref$dat.combat[,s] - site.out$dat.combat)))
  perror_ref <- c(perror_ref,
              max(c(abs(com_out_ref$dat.combat[,s] - site.out$dat.combat)/
                      site.out$dat.combat)))
}

#### Display errors ####
error # Maximum difference by batch
perror # Maximum difference as percent by batch

error_ref # Maximum difference by batch
perror_ref # Maximum difference as percent by batch
